src/polynomial/poly-grads.jl

```diff
@@ -0,0 +1,385 @@
+
+    (s0, kap0, kite_area2, kck), ∇s0 = compute_uniform_grad(b, r; b2, r2, sqarea, fourbrinv)
+
+    flux = ld.g_n[begin] * s0
+    ∇flux = ld.g_n[begin] * ∇s0
+    dfdg[begin] = s0
+
+    if ld.n_max == 0
+        dfdg[begin] -= flux * ld.norm * π
+        dfdb, dfdr = ∇flux * ld.norm
+        return flux * ld.norm, dfdg * ld.norm, dfdb, dfdr
+    end
+
+    ## compute linear term
+    (s1, Eofk, Em1mKdm), ∇s1 = compute_linear_grad(b, r; k2, kc, kc2, r2, b2, br, fourbr, sqbr, onembmr2, onembpr2, onembmr2inv, sqonembmr2)
+
+    flux += ld.g_n[begin + 1] * s1
+    ∇flux = ∇flux + ld.g_n[begin + 1] * ∇s1
+    dfdg[begin + 1] = s1
+
+    if ld.n_max == 1
+        dfdg[begin] -= flux * ld.norm * π
+        dfdg[begin + 1] -= flux * ld.norm * π * 2/3
+        dfdb, dfdr = ∇flux * ld.norm
+        return flux * ld.norm, dfdg * ld.norm, dfdb, dfdr
+    end
+
+    ## calculate quadratic term
+    s2, ∇s2 = compute_quadratic_grad(b, r; s0, r2, b2, kap0, kite_area2, k2, ∇s0)
+    flux += ld.g_n[begin + 2] * s2
+    ∇flux = ∇flux + ld.g_n[begin + 2] * ∇s2
+    dfdg[begin + 2] = s2
+    if ld.n_max == 2
+        dfdg[begin] -= flux * ld.norm * π
+        dfdg[begin + 1] -= flux * ld.norm * π * 2/3
+        dfdb, dfdr = ∇flux * ld.norm
+        return flux * ld.norm, dfdg * ld.norm, dfdb, dfdr
+    end
+
+    ## higher order (N > 3) terms require solving Mn and Nn integrals
+    if k2 < 0.5 && ld.n_max > 3
+        downwardM!(ld.Mn, ld.Mn_coeff; n_max=ld.n_max, sqonembmr2, sqbr, onemr2mb2, k, k2, sqarea, kap0, Eofk, Em1mKdm, kite_area2)
+        if b < bcut
+            Nn_lower!(ld.Nn, ld.Nn_coeff; Mn=ld.Mn, n_max=ld.n_max, kap0, kc, k2, sqbr, k, Eofk, Em1mKdm)
+        end
+    else
+        upwardM!(ld.Mn; sqbr, n_max=ld.n_max, sqonembmr2, onemr2mb2, sqarea, k2, kap0, Eofk, Em1mKdm, kite_area2)
+        if b < bcut
+            Nn_raise!(ld.Nn, ld.Mn; n_max=ld.n_max, onembpr2, kap0, kc, k2, sqbr, k, Eofk, Em1mKdm)
+        end
+    end
+
+    # compute remaining terms
+    for n in 3:ld.n_max
+        pofgn_M = 2 * r2 * ld.Mn[begin + n] - n / (n + 2) *
+                       (onemr2mb2 * ld.Mn[begin + n] + sqarea * ld.Mn[begin + n - 2])
+        flux -= ld.g_n[begin + n] * pofgn_M
+        dfdg[begin + n] = -pofgn_M
+        dpdr_M = 2 * r * ((n + 2) * ld.Mn[begin + n] - n * ld.Mn[begin + n - 2])
+
+        if b < bcut
+            dpdb_M = n * (ld.Mn[begin + n - 2] * (2 * r^3 + b^3 - b - 3 * r2 * b) + b * ld.Mn[begin + n] - 4 * r^3 * ld.Nn[begin + n - 2])
+        else
+            dpdb_M = n/b * ((ld.Mn[begin + n] - ld.Mn[begin + n - 2]) * (r2 + b2) + (b2 - r2)^2 * ld.Mn[begin + n - 2])
+        end
+        ∇flux = ∇flux - ld.g_n[begin + n] * SA[dpdb_M, dpdr_M]
+
+    end
+    dfdg[begin] -= flux * ld.norm * π
+    dfdg[begin + 1] -= flux * ld.norm * π * 2/3
+    dfdb, dfdr = ∇flux * ld.norm
+    return flux * ld.norm, dfdg * ld.norm, dfdb, dfdr
+end
+
+function frule((_, Δld, Δb, Δr), ::typeof(compute), ld::LD, b, r) where {LD<:PolynomialLimbDark}
```

Comment:
Am I missing something ir LD here and in most other methods below isn't actually used in the body of the function?

Reply:
https://julialang.slack.com/archives/C6G240ENA/p1617498835147700